src/mbgl/style/expression/distance.cpp

```diff
@@ -137,25 +190,87 @@ double pointsToPointsDistance(const mapbox::geometry::multi_point<double>& point
+                             const IndexRange& range,
+                             const mapbox::geometry::polygon<double>& polygon,
+                             mapbox::cheap_ruler::CheapRuler& ruler) {
+    if (!isRangeSafe(range, line.size())) {
+        mbgl::Log::Error(mbgl::Event::General, "index is out of range");
+        return InvalidDistance;
+    }
+    for (std::size_t i = range.first; i <= range.second; ++i) {
+        if (pointWithinPolygon(line[i], polygon, true /*trueOnBoundary*/)) return 0.0;
+    }
+
+    double dist = InfiniteDistance;
+    for (std::size_t i = range.first; i < range.second; ++i) {
+        const auto& p1 = line[i];
+        const auto& p2 = line[i + 1];
+        for (const auto& ring : polygon) {
+            for (std::size_t j = 0; j < ring.size() - 2; ++j) {
+                const auto& q1 = ring[j];
+                const auto& q2 = ring[j + 1];
+                if (segmentIntersectSegment(p1, p2, q1, q2)) return 0.0;
+                dist = std::min(dist, segmentToSegmentDistance(p1, p2, q1, q2, ruler));
+            }
         }
-        auto size1 = size / 2;
-        IndexRange range1(range.first, range.first + size1);
-        IndexRange range2(range.first + size1, range.second);
-        return std::make_pair(std::move(range1), std::move(range2));
-    } else {
-        if (size == 1) {
-            return std::make_pair(range, nullopt);
+    }
+    return dist;
+}
+
+double polygonToPolygonDistance(const mapbox::geometry::polygon<double>& polygon1,
+                                const mapbox::geometry::polygon<double>& polygon2,
+                                mapbox::cheap_ruler::CheapRuler& ruler,
+                                double currentMiniDist = InfiniteDistance) {
+    const auto bbox1 = getBBox(polygon1);
+    const auto bbox2 = getBBox(polygon2);
+    if (currentMiniDist != InfiniteDistance && bboxToBBoxDistance(bbox1, bbox2, ruler) >= currentMiniDist) {
+        return currentMiniDist;
+    }
+    const auto polygonIntersect = [](const mapbox::geometry::polygon<double>& poly1,
+                                     const mapbox::geometry::polygon<double>& poly2) {
+        for (const auto& ring : poly1) {
+            for (std::size_t i = 0; i <= ring.size() - 2; ++i) {
+                if (pointWithinPolygon(ring[i], poly2, true /*trueOnBoundary*/)) return true;
+            }
         }
-        auto size1 = size / 2 - 1;
-        IndexRange range1(range.first, range.first + size1);
-        IndexRange range2(range.first + size1 + 1, range.second);
-        return std::make_pair(std::move(range1), std::move(range2));
+        return false;
+    };
+    if (boxWithinBox(bbox1, bbox2)) {
+        if (polygonIntersect(polygon1, polygon2)) return 0.0;
+    } else if (polygonIntersect(polygon2, polygon1)) {
+        return 0.0;
     }
+
+    double dist = InfiniteDistance;
+    for (const auto& ring1 : polygon1) {
+        for (std::size_t i = 0; i < ring1.size() - 2; ++i) {
```

Comment:
Would this work with e.g. two triangles that both have 3 points and 3 segments?

Reply:
It depends on how we define the data for the polygon, normally I think the first and the last point should be the same, but you are right, in case polygon ring is just a container, we have to count all of the points. Anyway, it is the safest way to count the last point, I will update this.

Reply:
even if the ring is closed (ie. triangles would have 4 points) this would only go through first two segments: p0 - p1 and p1 - p2.

Reply:
You are right, it is an error :)